src/Pulsar.Client/Internal/DeadLetters.fs

```diff
@@ -0,0 +1,78 @@
+﻿namespace Pulsar.Client.Internal
+
+open System
+open Pulsar.Client.Api
+open Pulsar.Client.Common
+open System.Collections.Generic
+open FSharp.UMX
+open System.Threading.Tasks
+open FSharp.Control.Tasks
+open Microsoft.Extensions.Logging
+
+type internal DeadLettersProcessor
+    (policy: DeadLettersPolicy,
+     getTopicName: unit -> string,
+     getSubscriptionNameName: unit -> string,
+     createProducer: string -> Task<IProducer>) =
+
+    let store = Dictionary<MessageId, ResizeArray<Message>>()
+
+    let topicName =
+        if String.IsNullOrEmpty(policy.DeadLetterTopic) |> not then
+            policy.DeadLetterTopic
+        else
+            (sprintf "%s-%s-DLQ" (getTopicName()) (getSubscriptionNameName()))
+
+    let producer = lazy (
+
+        createProducer topicName
+    )
+
+    let sendMessageAsync (builder : MessageBuilder) messageId  =
+        task {
+            try
+                let! p = producer.Value
+                do! (p.SendAsync(builder) :> Task)
+                return true
+            with
+            | ex ->
+                Log.Logger.LogError(
+                    ex,
+                    "Send to dead letter topic exception with topic: {0}, messageId: {1}",
+                    topicName,
+                    messageId)
+                return false
+        } |> Async.AwaitTask
+
+    interface IDeadLettersProcessor with
+        member __.ClearMessages() =
+            store.Clear()
+
+        member __.AddMessage (message : Message) =
+            let id = message.MessageId
+            if message.RedeliveryCount |> int >= policy.MaxRedeliveryCount then
+                if store.ContainsKey(id) |> not then store.[id] <- ResizeArray<Message>()
+                store.[id].Add(message)
+
+        member __.RemoveMessage messageId =
+            store.Remove(messageId) |> ignore
+
+        member __.ProcessMessages messageId =
+            if  store.ContainsKey messageId then
+                store.[messageId]
+                |> Seq.map (fun m -> MessageBuilder(m.Payload, %m.MessageKey, m.Properties))
+                |> Seq.map (fun builder -> sendMessageAsync builder messageId)
+                |> Async.Parallel
+                |> Async.RunSynchronously
```

Comment:
Two issues
1) Java version doesn't send in parallel, let's stick to it
2) RunSynchronously blocks thread, please make ProcessMessages async

Reply:
I think code should block this thread because code should wait send result and return true or false...? All messages that are bundled under same message id (batch) should be sent. If one of sending attempt throws an exception function returns false. So, I think it's necessarily wait send result.

Reply:
I agree that result is needed, still there is no need to block thread for it :) Just make the method async and await it on the caller side

Reply:
Ok. Now dead letters processed in series and await moved on the caller side.

Reply:
Refactored it, please take a look what I wanted ;)

Reply:
Ok. I saw it. Now the idea is clear for me. I think there is small difference. As result you moved async await farther into MailboxAgent branch. But end effect will be the same I think.